### PR TITLE
Move BankHashStats from account's db to bank 

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

We keep track of bank_hash_stats for each bank. Currently, this stats live inside a locked hash map in Accounts DB. This cause a lock contention for every store of an accounts and hurts performance.

We want to get rid of the lock in the store code path.

#### Summary of Changes

Move bank_hash_stats from accounts db to bank.
Introduce AtomicBankHashStats to accumulate the stats in the bank.

Fixes #

### CODE REVIEW

File: runtime/src/bank.rs [hunk 1]
```diff
@@ -5101,7 +5110,9 @@ impl Bank {
                 )
             })
         });
-        self.rc.accounts.store_accounts_cached(accounts);
+        self.rc
+            .accounts
+            .store_accounts_cached(accounts, Some(&self.bank_hash_stats));
```
Comment (brooksprumo) [inline runtime/src/bank.rs]: Instead of passing this new bank_hash_stats down to AccountsDb, can we instead just update the stats here inside Bank?

We know whenever we call Accounts/AccountsDb store_accounts, so I'm hoping that's possible. Wdyt?

File: runtime/src/serde_snapshot.rs [hunk 2]
```diff
@@ -1209,12 +1209,6 @@ where
         old_accounts_delta_hash.is_none(),
         "There should not already be an AccountsDeltaHash at slot {snapshot_slot}: {old_accounts_delta_hash:?}",
         );
-    let old_stats = accounts_db
-        .update_bank_hash_stats_from_snapshot(snapshot_slot, snapshot_bank_hash_info.stats);
```
Comment (brooksprumo) [inline runtime/src/serde_snapshot.rs:1213]: When loading from a snapshot, we should still set the Bank's `bank_hash_stats` appropriately.

Comment (HaoranYi) [reply runtime/src/serde_snapshot.rs:1213]: fixed in [019ffd07c1b9c566b0449a244682a153d2adc278](https://github.com/anza-xyz/agave/pull/3821/commits/019ffd07c1b9c566b0449a244682a153d2adc278)

File: runtime/src/bank.rs (hunk 1 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: yeah. updated!

File: accounts-db/src/accounts_db/tests.rs [hunk 3]
```diff
@@ -2411,13 +2409,6 @@ fn test_bank_hash_stats() {
     account.set_executable(true);
     db.store_for_tests(some_slot, &[(&key, &account)]);
     db.add_root(some_slot);
-
-    let stats = db.get_bank_hash_stats(some_slot).unwrap();
-    assert_eq!(stats.num_updated_accounts, 1);
-    assert_eq!(stats.num_removed_accounts, 1);
-    assert_eq!(stats.num_lamports_stored, 1);
-    assert_eq!(stats.total_data_len, 2 * some_data_len as u64);
-    assert_eq!(stats.num_executable_accounts, 1);
```
Comment (brooksprumo) [inline accounts-db/src/accounts_db/tests.rs:2420]: Should this whole test be removed?

File: runtime/src/snapshot_package.rs [hunk 4]
```diff
@@ -79,7 +79,7 @@ impl AccountsPackage {
             // that guarantees this slot will have an accounts delta hash.
             let accounts_delta_hash = accounts_db.get_accounts_delta_hash(slot).unwrap();
             // SAFETY: Every slot *must* have a BankHashStats entry in AccountsDb.
-            let bank_hash_stats = accounts_db.get_bank_hash_stats(slot).unwrap();
+            let bank_hash_stats = bank.get_bank_hash_stats();
```
Comment (brooksprumo) [inline runtime/src/snapshot_package.rs:81]: The comment above this can be removed. We are no longer querying AccountsDb, and no longer have an unwrap.

File: runtime/src/bank.rs [hunk 5]
```diff
@@ -1337,6 +1384,7 @@ impl Bank {
             cache_for_accounts_lt_hash: RwLock::new(AHashMap::new()),
             stats_for_accounts_lt_hash: AccountsLtHashStats::default(),
             block_id: RwLock::new(None),
+            bank_hash_stats: AtomicBankHashStats::new(&parent.bank_hash_stats.load()),
```
Comment (brooksprumo) [inline runtime/src/bank.rs]: This should be values unique to this slot, not `parent`.

```suggestion
            bank_hash_stats: AtomicBankHashStats::default(),
```

File: runtime/src/bank.rs [hunk 6]
```diff
@@ -1749,6 +1797,7 @@ impl Bank {
             cache_for_accounts_lt_hash: RwLock::new(AHashMap::new()),
             stats_for_accounts_lt_hash: AccountsLtHashStats::default(),
             block_id: RwLock::new(None),
+            bank_hash_stats: AtomicBankHashStats::default(),
```
Comment (brooksprumo) [inline runtime/src/bank.rs]: We should get these values from the snapshot (fields).

Oh, never mind, I see that we set the value after new_from_fields() runs. Can you add a comment here that indicates that?

Or maybe another option that I think I like more is to add the BankHashStats to BankFieldsToDeserialize, and then we (1) don't need to set the value after the fact, and (2) don't need to add another parameter to this function.

File: runtime/src/serde_snapshot.rs [hunk 7]
```diff
@@ -871,6 +885,7 @@ where
             .map(|bank_fields| bank_fields.capitalization),
     );
     let bank_fields = bank_fields.collapse_into();
+    let bank_hash_stats = snapshot_accounts_db_fields.extract_bank_hash_stats();
```
Comment (brooksprumo) [inline runtime/src/serde_snapshot.rs]: An alternative is to use `reconstruct_accountsdb_from_fields()` and return the stats as part of `ReconstructedAccountsDbInfo`, since it already done `collapse_into()`, which is basically the same impl as `extract_bank_hash_stats()`. 

I can see pros/cons to each. I kind of like avoiding duplicating the code that's already in`collapse_into()` though.

File: runtime/src/bank.rs [hunk 8]
```diff
@@ -1011,6 +1015,49 @@ pub struct ProcessedTransactionCounts {
...
+    pub fn load(&self) -> BankHashStats {
+        BankHashStats {
+            num_updated_accounts: self.num_updated_accounts.load(Relaxed),
+            num_removed_accounts: self.num_removed_accounts.load(Relaxed),
+            num_lamports_stored: self.num_lamports_stored.load(Relaxed),
+            total_data_len: self.total_data_len.load(Relaxed),
+            num_executable_accounts: self.num_executable_accounts.load(Relaxed),
+        }
+    }
+}
```
Comment (brooksprumo) [inline runtime/src/bank.rs:1098]: I would prefer if this struct/impl lived next to BankHashStats. Maybe BankHashStats can be moved here? Or AtomicBankHashStats can go into accounts_db/stats.rs?

It could also be a follow up PR.

File: runtime/src/bank.rs (hunk 5 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: I thought that we are accumulating the stats on the fork. 
So child bank should start with the stat from the parent bank, right?

Comment (brooksprumo) [reply runtime/src/bank.rs]: No, it's the stats of the single slot.

File: accounts-db/src/accounts_db/tests.rs (hunk 3 above)
Comment (HaoranYi) [reply accounts-db/src/accounts_db/tests.rs:2420]: yes. 
https://github.com/anza-xyz/agave/pull/3821/commits/379309ed7d25b8602f7a2e5f83d2ea8c3847f552

File: runtime/src/snapshot_package.rs (hunk 4 above)
Comment (HaoranYi) [reply runtime/src/snapshot_package.rs:81]: done in 
https://github.com/anza-xyz/agave/pull/3821/commits/d1ce9874afeeae0b64d6dc44f933809f07632c3e

File: runtime/src/bank.rs (hunk 8 above)
Comment (HaoranYi) [reply runtime/src/bank.rs:1098]: done in 
https://github.com/anza-xyz/agave/pull/3821/commits/b9f2c914cc0e1fbd56b426d2be57cf0649dbe8b5

File: runtime/src/serde_snapshot.rs (hunk 7 above)
Comment (HaoranYi) [reply runtime/src/serde_snapshot.rs]: done in 
https://github.com/anza-xyz/agave/pull/3821/commits/863005e373478bd338b749531bafb9267caeff13

File: runtime/src/bank.rs [hunk 9]
```diff
@@ -2786,9 +2874,9 @@ impl Bank {
         );
         assert!(!self.freeze_started());
         thread_pool.install(|| {
-            stake_rewards
-                .par_chunks(512)
-                .for_each(|chunk| self.rc.accounts.store_accounts_cached((slot, chunk)))
+            stake_rewards.par_chunks(512).for_each(|chunk| {
+                self.rc.accounts.store_accounts_cached((slot, chunk));
+            })
```
Comment (brooksprumo) [inline runtime/src/bank.rs]: This formatting change look unrelated to the PR. Can it be reverted?

File: runtime/src/serde_snapshot.rs (hunk 7 above)
Comment (brooksprumo) [reply runtime/src/serde_snapshot.rs]: Can BankHashStats be put in ReconstructedAccountsDbInfo?

File: runtime/src/bank.rs (hunk 9 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: done in
https://github.com/anza-xyz/agave/pull/3821/commits/150d95f0e33540711b89d53ecd593a59a1129224

File: runtime/src/bank.rs (hunk 5 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: ah. done in 
https://github.com/anza-xyz/agave/pull/3821/commits/34074bab8790780610414c73c792394ca4f62834

**Review by brooksprumo — COMMENTED** [review]
Looks like we're missing updating the slot's bank_hash_stats in two other places:
1. store_stake_accounts()
2. commit_transactions()

Both of these functions call into accounts/accountsdb directly, so they bypass the Bank::store_accounts() logic.

File: runtime/src/serde_snapshot.rs (hunk 7 above)
Comment (HaoranYi) [reply runtime/src/serde_snapshot.rs]: done in https://github.com/anza-xyz/agave/pull/3821/commits/383ab3e8a63273f44cd0a1ccf6b139b6996923c4

**Comment by HaoranYi** [comment]
> Looks like we're missing updating the slot's bank_hash_stats in two other places:
> 
> 1. store_stake_accounts()
> 2. commit_transactions()
> 
> Both of these functions call into accounts/accountsdb directly, so they bypass the Bank::store_accounts() logic.

done in https://github.com/anza-xyz/agave/pull/3821/commits/284e7384d8a1a903d03b76b1ce7800386114115e

File: runtime/src/bank.rs (hunk 6 above)
Comment (brooksprumo) [reply runtime/src/bank.rs]: I see that the current impl now passes BankHashStats as a fn parameter to Bank::new_from_fields(). Can it instead be added to BankFieldsToDeserialize, and then no new fn parameters will be needed.

File: runtime/src/bank.rs [hunk 10]
```diff
@@ -7112,6 +7211,14 @@ impl Bank {
...
             .add_builtin(self, program_id, name, builtin)
     }
+
+    pub fn get_bank_hash_stats(&self) -> BankHashStats {
+        self.bank_hash_stats.load()
+    }
+
+    pub fn set_bank_hash_stats(&mut self, bank_hash_stats: &BankHashStats) {
+        self.bank_hash_stats = AtomicBankHashStats::new(bank_hash_stats);
+    }
```
Comment (brooksprumo) [inline runtime/src/bank.rs]: Can this fn be removed now? I don't think there are any more callers.

**Comment by HaoranYi** [comment]
The log for BankHashStats matched with master when running on mainnet.

- this pr

```
[2024-12-03T14:34:51.113908867Z INFO  solana_runtime::bank] bank frozen: 305167208 hash: 4qzkBXbkow8KTSQ3gpgwL7oGEQVmHkMFyxRJ2qSjzF63 accounts_delta: W2cZQQKRPkkx5gGaHjMKpTCvz6SNAjnLMKw2tXtmqoG signature_count: 2131 last_blockhash: CRSrNfrT2CMC32SxTZpXs65yx4ShaxyRHo9NASL3Ew93 capitalization: 589423807506551765, stats: BankHashStats { num_updated_accounts: 6359, num_removed_accounts: 233, num_lamports_stored: 17440095079623122, total_data_len: 39964086, num_executable_accounts: 3 }
[2024-12-03T14:34:51.533962317Z INFO  solana_runtime::bank] bank frozen: 305167209 hash: 3JFoM5ceBXWhrL41Hk2igBRe4s8Qq3JALFRGdGA77VJh accounts_delta: GyymD8hD8PYWiLtzUXZNbamY77pMzqYiNVNZYk3JPrg3 signature_count: 1136 last_blockhash: J9U3bY8CwwotGynEazzEN7MUC5vnsa15AgepuGY74sbe capitalization: 589423807419181024, stats: BankHashStats { num_updated_accounts: 3790, num_removed_accounts: 113, num_lamports_stored: 6829130698133182, total_data_len: 27757127, num_executable_accounts: 5 }

```

- master
```
[2024-12-03T14:34:50.954047096Z INFO  solana_runtime::bank] bank frozen: 305167208 hash: 4qzkBXbkow8KTSQ3gpgwL7oGEQVmHkMFyxRJ2qSjzF63 accounts_delta: W2cZQQKRPkkx5gGaHjMKpTCvz6SNAjnLMKw2tXtmqoG signature_count: 2131 last_blockhash: CRSrNfrT2CMC32SxTZpXs65yx4ShaxyRHo9NASL3Ew93 capitalization: 589423807506551765, stats: BankHashStats { num_updated_accounts: 6359, num_removed_accounts: 233, num_lamports_stored: 17440095079623122, total_data_len: 39964086, num_executable_accounts: 3 }
[2024-12-03T14:34:51.365612394Z INFO  solana_runtime::bank] bank frozen: 305167209 hash: 3JFoM5ceBXWhrL41Hk2igBRe4s8Qq3JALFRGdGA77VJh accounts_delta: GyymD8hD8PYWiLtzUXZNbamY77pMzqYiNVNZYk3JPrg3 signature_count: 1136 last_blockhash: J9U3bY8CwwotGynEazzEN7MUC5vnsa15AgepuGY74sbe capitalization: 589423807419181024, stats: BankHashStats { num_updated_accounts: 3790, num_removed_accounts: 113, num_lamports_stored: 6829130698133182, total_data_len: 27757127, num_executable_accounts: 5 }

```

**Review by brooksprumo — COMMENTED** [review]
I think the logic looks correct now. We should test this against another node on mnb to ensure both produce the same bank hash stats, as these numbers are used to debug consensus mismatches.

Can you run master vs this PR and ensure the bank hash details that are logged in the "bank frozen" line are always the same? Ideally over an epoch boundary, since that'll exercise also storing the stake accounts.

Edit: Nice! I see you just did this test: https://github.com/anza-xyz/agave/pull/3821#issuecomment-2514757701

File: runtime/src/bank.rs (hunk 6 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: done in https://github.com/anza-xyz/agave/pull/3821/commits/0823de281af77dd54fa34e588cdb69bee75e4a8f

File: runtime/src/bank.rs (hunk 10 above)
Comment (HaoranYi) [reply runtime/src/bank.rs]: done in https://github.com/anza-xyz/agave/pull/3821/commits/fd6e1b6bc71921823306311d5c532d384d41d2ca

**Comment by HaoranYi** [comment]
Yes. I will keep it running and check for the slot at the next epoch boundary, which is about 1d7h from now.

**Review by brooksprumo — APPROVED** [review]
:shipit: 

Looks good to me! Probably best to wait and confirm the stats match master across an epoch boundary.

File: runtime/src/bank.rs [hunk 11]
```diff
@@ -1011,6 +1016,87 @@ pub struct ProcessedTransactionCounts {
...
+/// Account stats for computing the bank hash
+/// This struct is serialized and stored in the snapshot.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+#[derive(Clone, Default, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct BankHashStats {
+    pub num_updated_accounts: u64,
+    pub num_removed_accounts: u64,
+    pub num_lamports_stored: u64,
+    pub total_data_len: u64,
+    pub num_executable_accounts: u64,
```
Comment (jeffwashington) [inline runtime/src/bank.rs:1028]: i'm struggling to see where these are reported. What am I missing?

Comment (HaoranYi) [reply runtime/src/bank.rs:1028]: it is not reported in metric. We logged it when we freeze the bank.

https://github.com/anza-xyz/agave/blob/df27fb3a7386f0d2cc64df81c46a09ab79b18ba0/runtime/src/bank.rs#L5644

seems to only be used for debugging consensus issue.

Comment (jeffwashington) [reply runtime/src/bank.rs:1028]: ah. I was looking for a use of the specific fields. I didn't realize the entire thing was debug logged.

**Comment by HaoranYi** [comment]
stats matched at the epoch boundary. 

```
[2024-12-04T20:48:22.626606613Z INFO  solana_runtime::bank] bank frozen: 305424000 hash: GG83exajA53vbdsA8DPxougAybY6GPGsdT6eXkP4nFyT accounts_delta: H9uhPuch44MGrmzFUoHg4eBnZbF6pMbwhQF5nt84VCbb signature_count: 0 last_blockhash: oJ2wJEUjva6BU5uapEqDDZcnTwBEn1uRnx6vnLG56PN capitalization: 589453979955825477, stats: BankHashStats { num_updated_accounts: 1431, num_removed_accounts: 0, num_lamports_stored: 394203918046202, total_data_len: 5531172, num_executable_accounts: 0 }
[2024-12-04T20:48:25.022027026Z INFO  solana_runtime::bank] bank frozen: 305424001 hash: 5jt3dwBfJuYksn4UFctAjAUNwL9xwCbb3yiXmhijgd9 accounts_delta: FEJAMNFYjywEyqBtFkLcYkzDL6Jj5SwUNHbnaEFSsm78 signature_count: 28 last_blockhash: 7kxx1ChdRRVno6K5Vn1ANNAm2i8GvaBZa446HW3HMJXo capitalization: 589454371122902139, stats: BankHashStats { num_updated_accounts: 4311, num_removed_accounts: 5, num_lamports_stored: 1686863971165642, total_data_len: 1822705, num_executable_accounts: 0 }

```

```
[2024-12-04T20:48:22.385496541Z INFO  solana_runtime::bank] bank frozen: 305424000 hash: GG83exajA53vbdsA8DPxougAybY6GPGsdT6eXkP4nFyT accounts_delta: H9uhPuch44MGrmzFUoHg4eBnZbF6pMbwhQF5nt84VCbb signature_count: 0 last_blockhash: oJ2wJEUjva6BU5uapEqDDZcnTwBEn1uRnx6vnLG56PN capitalization: 589453979955825477, stats: BankHashStats { num_updated_accounts: 1431, num_removed_accounts: 0, num_lamports_stored: 394203918046202, total_data_len: 5531172, num_executable_accounts: 0 }
[2024-12-04T20:48:24.667013810Z INFO  solana_runtime::bank] bank frozen: 305424001 hash: 5jt3dwBfJuYksn4UFctAjAUNwL9xwCbb3yiXmhijgd9 accounts_delta: FEJAMNFYjywEyqBtFkLcYkzDL6Jj5SwUNHbnaEFSsm78 signature_count: 28 last_blockhash: 7kxx1ChdRRVno6K5Vn1ANNAm2i8GvaBZa446HW3HMJXo capitalization: 589454371122902139, stats: BankHashStats { num_updated_accounts: 4311, num_removed_accounts: 5, num_lamports_stored: 1686863971165642, total_data_len: 1822705, num_executable_accounts: 0 }

```

**Review by jeffwashington — APPROVED** [review]
fine by me. only used for debugging.
